keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/notifications/PinnedNotification.kt

```diff
@@ -0,0 +1,85 @@
+package com.netflix.spinnaker.keel.slack.notifications
+
+import com.netflix.spinnaker.keel.api.ScmInfo
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.artifacts.getScmBaseLink
+import com.netflix.spinnaker.keel.events.SlackPinnedNotification
+import com.netflix.spinnaker.keel.slack.SlackNotifier
+import com.slack.api.model.kotlin_extension.block.withBlocks
+import org.apache.logging.log4j.util.Strings
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+
+@Component
+class PinnedNotification (
+  private val slackNotifier: SlackNotifier,
+  private val scmInfo: ScmInfo,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+){
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @EventListener(SlackPinnedNotification::class)
+  fun onPinnedEvent(event: SlackPinnedNotification) {
+    log.debug("Sending pinnedEnvironment notification for application ${event.application}")
+
+    with(event) {
+      val env = Strings.toRootUpperCase(pin.targetEnvironment)
+      val pinnedArtifactUrl = "$spinnakerBaseUrl/#/applications/${application}/environments/${pinnedArtifact?.reference}/${pinnedArtifact?.version}"
+      val repoLink = generateRepoLink(currentArtifact?.gitMetadata)
+      var details = ""
+      val blocks = withBlocks {
+        header {
+          text("$env is pinned", emoji = true)
+        }
+
+        section {
+          markdownText("*Version:* ~#${event.currentArtifact?.buildMetadata?.number}~ → <$pinnedArtifactUrl|#${pinnedArtifact?.buildMetadata?.number}> " +
+            "by ${pinnedArtifact?.gitMetadata?.author}\n " +
+            "*Where:* $env\n\n " +
+            "${pinnedArtifact?.gitMetadata?.commitInfo?.message}")
+          accessory {
+            image(imageUrl = "https://raw.githubusercontent.com/gcomstock/managed.delivery/master/src/icons/pinned.png", altText = "pinned")
```

Comment:
My instinct is that we should avoid hotlinking to GitHub like this, but I don't know what the common pattern is for linking images in Slack notifications.

Reply:
good point, let's ask @erikmunson :)

Reply:
We could hotlink to Spinnaker UI assets from deck instead. 🙂

Reply:
I originally thought the same Luis, but then later discovered that the public assets (i.e. available without auth) we serve from deck are all behind the Netflix internal VPN. That would mean that notifications would look broken for anyone on the public internet.

I'd love it if we could figure out a way to expose stuff outside the Netflix VPN without jumping through too many hoops, but also haven't had the time / mental focus to go any further looking into it. If anybody has thoughts on it please raise them so we can discuss!

For now I think having something that's hacky but publicly accessible is good enough to ship v0 of this feature, given its level of importance. Still worth improving it though.

Reply:
Ah, right, doh! [spinnaker.io](https://spinnaker.io)? 🙂